### ISSUE TITLE
Reasonable default value for `udp_max_queries`

### ISSUE THREAD

**Issue by Stevenjin8** [c-ares/c-ares#913]
Hello there,

At Istio, we are facing some issues with c-ares: https://github.com/istio/istio/issues/53577.
Our proposed solution is to set `udp_max_queries` to some nonzero value.
We set a default value for `udp_max_queries` instead of providing a configuration option as to minimize our API surface.

Is there guidance on what a good default would be?

**Comment by bradh352** [comment]
I read through that and some of the other linked stuff, interesting. 

Honestly `udp_max_queries` was a security toggle since the ephemeral port is part of the entropy involved to prevent cache poisoning attacks, not meant for this use-case.

I think the risk of using udp_max_queries when there are a LOT of dns requests is possibly running out of ephemeral ports and all the necessary overhead of "connection" tracking in the iptables nat methodology that is used.  I'm not sure I understand the failed server/pod aspect fully.

There's typically about 15000 ephemeral ports available per pair of endpoints (granted you can change this via `net.ipv4.ip_local_port_range` on linux), so you really want to make sure you don't blow past that.  A maximum query timeout is roughly 5s (though c-ares these days auto-adjusts based on network conditions and is likely lower).  So the maximum number of queries you can perform per second (reliably) is something like:
`net.ipv4.ip_local_port_range / 5 * udp_max_queries`.  If you have an upper threshold that you know you need to support, you can probably calculate that backwards.

Then you have the netfilter connection tracking aspect, which will cache udp "connections" for 30s by default: 
`net.netfilter.nf_conntrack_udp_timeout`.  Not only is it a lot to track, but if you're cycling through ephemeral ports really quickly, you might actually get a duplicate "stale" port from the table which if I'm reading it correctly might route you back to the wrong pod which is really down.  So its possible the limit really should be calculated like `net.ipv4.ip_local_port_range / net.netfilter.nf_conntrack_udp_timeout * udp_max_queries`.  So that means at default values for those settings, and I see in the ticket you referenced says 15K req/s, that would mean a value of `30` might be appropriate.

But ... now coming back around to a failed server/pod, what happens with UDP in this case with the iptables NAT methodology?  Is this still a concern with current c-ares releases?  I would hope that maybe ICMP unreachable packets would be generated which as of #821 would actually terminate the udp "connection" (this is a fairly recent change as of c-ares 1.32.3).  But if that's not happening, and you happen to be at a very slow DNS lookup time or `udp_max_queries` is set very high, you'd get resolution issues there until the max was reached which who knows how "long" that might take.

**Comment by bradh352** [comment]
So, now that my previous reply has you thinking `udp_max_queries` isn't the right way to go ... 

I don't suppose there's a way to somehow either get the system to assign multiple ips or multiple ports to the dns service is there?  If you could get c-ares to see multiple servers I think it would solve a lot of issues, even if it really just speaks to the same cluster of pods.   You could use something like ARES_OPT_ROTATE to make sure traffic gets distributed rather than all packets going to a single node for the 'connection lifetime'.  It might also make sense to use TCP as these days we support TCP FastOpen which gets rid of most of the downsides to using TCP with DNS. https://c-ares.org/features/#tcp-fastopen-0-rtt

If that's not an option, then it may be best to enhance c-ares for this specific use-case.  We could create an feature to force c-ares to establish multiple configurable "connections" to a single endpoint, which normally wouldn't make sense, but in this specific usecase we know it is balanced with multiple clustered nodes, so it would.  The connections would be used round-robin or least-connections. We'd still need to consider the failed pod aspect, but I'd need more information on that ... how to detect, etc.

**Comment by Stevenjin8** [comment]
Just to make sure I understand, `udp_max_queries` sets the number of dns queries a udp socket can make before ares switches to another one. Because netfilter tracks udp "connections", and there are a finite number of ports, setting udp_max_queries effectively sets an upper bound for how many dns request ARES can make every second. 
Thus, if we set the value too low, we will throttle dns resolutions. If we set it too high, ares might get stuck with a bad "connection" for longer. And maybe #821 fixes this issue altogether.

Summarizing your second comment, there might be other ways to solve the issue. Namely, using TCP or setting ARES_OPT_ROTATE. I'm hesitant to do these because we can't guarantee that pods have TCP fast open enabled and Envoy doesn't expose ARES_OPT_ROTATE (I guess I could add it, but would rather not since we aren't even sure that this will solve the issue). I also think that enhancing c-ares might be an overkill, since this issue has only been reported once.

So my conclusion is to set this value to something fairly high (100-1000?) because we are pretty confident it will solve the issue and requires minimal changes. A high value ensures that we aren't throttling. It might cause Envoy to get "stuck" for longer, but its still better than the default of it getting stuck forever. For example, at 15k requests per second and a udp_max_queries of 1000, envoy would get unstuck in 1/15th of a second, which sounds fine to me. Please let me know if this makes sense.

**Comment by bradh352** [comment]
Yes, what you said sounds accurate.  Its just the solution won't really do proper load balancing across the coredns pods, it will slam each pod with a bunch of simultaneous queries before moving to the next rather than a more graceful distribution pattern.  Also all the adjacent requests sent to the prior pod will need to timeout and retry before getting unstuck unless #821 fixed that, not sure the usage pattern if that's a factor.

I'm always looking for the best solution to things though, but maybe for what you're doing its "good enough".

Regardless, you should probably push Envoy to adopt a newer c-ares version, I took a glance and it looks like their repo is forcing 1.21 which is missing a lot of enhancements and a minor security fix.

**Comment by bradh352** [comment]
I'm going to go ahead and close this issue.  Open another issue if you think it would be worth pursuing a c-ares flag that Envoy could adopt for their architecture.

**Comment by Stevenjin8** [comment]
@bradh352 thanks so much for your help!
